Upload archived resources to filestore

ismail-shahzad opened the pull request:
FIXES: https://gitlab.com/datopian/core/support/-/issues/149

Now all the archived resources can be uploaded to filestore.

## Description

Uploads the archived resources to s3filestore if CKANEXT__ARCHIVER__S3UPLOAD_ENABLE is set True in the .env file.

Adds two new functions

  - upload_archived_resource: Uploads the resources to s3filestore in directory <S3FILESTORE__AWS_BUCKET_NAME>/<S3FILESTORE__AWS_STORAGE_PATH>/archived_resources/resource_id/archive_resource_id
  - generate_cache_url: Generates a presigned url to download the resource from the s3filestore which expires after 1day(86400s) and returns that as cache_url
- Also removes archive_dir from public directory which was made public temporarirly for debugging.

ismail-shahzad committed:
[tasks.py] Adds code to upload the archived resources to s3filestore

Uploads the archived resources to s3filestore if CKANEXT__ARCHIVER__S3UPLOAD_ENABLE is set True in the .env file.

Adds two new functions

  - upload_archived_resource: Uploads the resources to s3filestore in directory <S3FILESTORE__AWS_BUCKET_NAME>/<S3FILESTORE__AWS_STORAGE_PATH>/archived_resources/
  - generate_cache_url: Generates a presigned url to download the resource from the s3filestore which expires after 1day(86400s) and returns that as cache_url

ismail-shahzad committed:
[tasks.py][s]: Fixes return type of upload_archive_resource

ismail-shahzad committed:
[plugin.py][s]: Removes archive_dir from public directory

ismail-shahzad committed:
[tasks.py][s]: Adds resource_dir in archived path

ismail-shahzad committed:
[tasks.py] Fixes return type of upload_archived_resource

zelima commented on ckanext/archiver/tasks.py:
```diff
@@ -548,11 +548,54 @@ def archive_resource(context, resource, log, result=None, url_timeout=30):
+    Uploads the resources to s3filestore in directory
+    <S3FILESTORE__AWS_BUCKET_NAME>/<S3FILESTORE__AWS_STORAGE_PATH>/archived_resources/
+    '''
+    storage_path = config.get('ckanext.s3filestore.aws_storage_path')
```
Seems like you are depended on `aws_storage_path`. Let's not just assume that it's set but a validation error if it's not. Or is it not required? If not required this will return `None` and `os.path.join` will error. in that case do `config.get('ckanext..', '')` so that you get empty string

ismail-shahzad replied:
It is required and is usually available for all the instances as it is used while resources are uploaded to a filestore by instances.
Will add validation error if not available in config.

zelima replied:
Let's be sure it's set or if not user knows it should be. This extension is used outside of our instances as well

ismail-shahzad replied:
Will add relevant info in the README

for storage path we can do something like

```
if not config.get('ckanext.s3filestore.aws_storage_path'):
        log.warning('Not saved to filestore because no value for '
                    'ckanext.s3filestore.aws_storage_path in config')
        raise ArchiveError(_('No value for ckanext.s3filestore.aws_storage_path in config'))
```

ismail-shahzad replied:
Implemented the above-mentioned logic

Added config settings. SEE Point no. 6 of README.
https://github.com/datopian/ckanext-archiver/blob/feature/filestore_resource_upload/README.rst#config-settings

zelima commented on ckanext/archiver/tasks.py:
```diff
@@ -548,11 +548,54 @@ def archive_resource(context, resource, log, result=None, url_timeout=30):
+    client = s3.client(service_name='s3', endpoint_url=host_name)
+    cache_url = client.generate_presigned_url(ClientMethod='get_object',
+                                              Params={'Bucket':bucket.name,'Key':key_path},
+                                              ExpiresIn=86400)
```
What is max possible `ExpiresIn`? Let's set it to max.

ismail-shahzad replied:
A presigned URL can be valid for a maximum of seven days because the signing key you use in signature calculation is valid for up to seven days.

zelima replied:
Great let's set to 7 days than

ismail-shahzad replied:
ok

ismail-shahzad replied:
`ExpiresIn` set to 7 days(604800s)

zelima reviewed (changes requested):
@MuhammadIsmailShahzad great work! Let's add a few words in README about `ckanext.archiver.s3upload_enable`...

Also that some corn jobs are needed to be setup and how.

ismail-shahzad committed:
[tasks.py] Adds a check for ckanext.s3filestore.aws_storage_path

ismail-shahzad committed:
[tasks.py][s]: Refactors logic and increases expiry

- Refactors the logic to check for storage_path
- Increases the expiry for presigned url from 1 to 7 days

ismail-shahzad committed:
[README.rst] Adds instructions to upload archived resources to s3filestore.

ismail-shahzad committed:
[REAME.rst][s]: Minor format fixes

ismail-shahzad committed:
[README.rst][s]: Minor typo fixes

zelima commented on README.rst:
```diff
@@ -270,6 +270,21 @@ Config settings
                 root /www/resource_cache;
             }
 
+6.  If you are not serving your archived resources through a webserver, you can upload them to ``s3filestore``. 
```
say

Alternatively, you can upload archived resources to cloud storage. In this case, make sure that you have ...

zelima commented on README.rst:
```diff
@@ -270,6 +270,21 @@ Config settings
+6.  If you are not serving your archived resources through a webserver, you can upload them to ``s3filestore``. 
+    Make sure you have properly setup `ckanext-s3filestore <https://github.com/datopian/ckanext-s3filestore>`_ for this.
+
+    Add the settings to the CKAN config file:
```
Add the following values to the CKAN config file:

zelima commented on README.rst:
```diff
@@ -270,6 +270,21 @@ Config settings
+
+    Add the settings to the CKAN config file:
+
+      * ``ckanext.archiver.s3upload_enable`` = ``True`` to enable upload to filestore. If unset or ``False``, the resources are going to be archived locally instead of on s3filestore.
```
``ckanext.archiver.s3upload_enable`` = ``True`` to enable upload to cloud storage, defaluts to `False`

zelima commented on README.rst:
```diff
@@ -270,6 +270,21 @@ Config settings
+    Add the settings to the CKAN config file:
+
+      * ``ckanext.archiver.s3upload_enable`` = ``True`` to enable upload to filestore. If unset or ``False``, the resources are going to be archived locally instead of on s3filestore.
+      * ``ckanext.s3filestore.aws_storage_path`` = ``my-site-name``. Your filestore project path. For example ckan/storage_path/archived_resource_dir. This must be set in order to upload files to filestore.
```
Say Required instead of This must be set in order to upload files to filestore.

zelima commented on README.rst:
```diff
@@ -270,6 +270,21 @@ Config settings
+      * ``ckanext.archiver.s3upload_enable`` = ``True`` to enable upload to filestore. If unset or ``False``, the resources are going to be archived locally instead of on s3filestore.
+      * ``ckanext.s3filestore.aws_storage_path`` = ``my-site-name``. Your filestore project path. For example ckan/storage_path/archived_resource_dir. This must be set in order to upload files to filestore.
+
+    The resources are uploaded to s3filestore in the directory ``s3filestore.aws_bucket_name/s3filestore.aws_storage_path/archived_resources/resource_id/``.
```
uploaded to cloud storage or S3 file storage (s3filestore is an extension name)

ismail-shahzad replied:
@zelima 

Everything in the code is done keeping s3filestore in mind

https://github.com/datopian/ckanext-archiver/blob/feature/filestore_resource_upload/ckanext/archiver/tasks.py#L565-L599

Would other cloud-storages be compatible with this?

zelima replied:
@MuhammadIsmailShahzad I don't know, will it?

zelima replied:
Speaking of that, don't we need updating requirements.txt?

ismail-shahzad replied:
It is only going to work for s3filestore. So not for all cloudstorages.

And we are adding no new packages or anything. Everything needed for filestore is being installed by s3filestore extension.

ismail-shahzad committed:
[README.rst] Fixes wording for adding s3filestore settings
